arch/arm/src/nrf52/nrf52_rtc_lowerhalf.h

```diff
@@ -0,0 +1,58 @@
+/****************************************************************************
+ * arch/arm/src/nrf52/nrf52_rtc_lowerhalf.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __ARCH_ARM_SRC_NRF52_NRF52_RTC_LOWERHALF_H
+#define __ARCH_ARM_SRC_NRF52_NRF52_RTC_LOWERHALF_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <nrf52_rtc.h>
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nrf52_rtc_initialize
+ *
+ * Description:
+ *   Bind the configuration timer to a timer lower half instance and
+ *   register the timer drivers at 'devpath'
+ *
+ * Input Parameters:
+ *   devpath - The full path to the timer device.  This should be of the
+ *     form /dev/timer0
+ *   timer - the timer's number.
+ *   timer_drvr - Pointer where to store resulting timer_lowerhalf_s*
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success; A negated errno value is returned
+ *   to indicate the nature of any failure.
+ *
+ ****************************************************************************/
+
+int nrf52_rtc_initialize(FAR const char *devpath, uint8_t timer,
+                         FAR struct timer_lowerhalf_s **timer_drvr);
```

Comment:
Does NFR52 RTC will clear the count when the count value equal comparison value? If the anwser isn't, it's better to export RTC through oneshot_lowerhalf_s interface, because the timestamp implemented through arch_timer(timer_lowerhalf_s) will be drift in the long run.

Reply:
No, RTC does not support auto-clearing. I'm doing this by manually clearing it on interrupt (see `nrf52_rtc_handler`). nRF52 has a peripheral which allows you to tie events to tasks, so you could connect the COMPARE event to the CLEAR task and then it will be taken care of by hardware. However, this uses up one PPI channel which may not be desireable. This is stated in the comment here (expand the _lowerhalf file diff): https://github.com/apache/incubator-nuttx/pull/1705/files#diff-57c967820db5f66391fba3ad66318487R156

Anyway, I also wondered what was the proper way to expose the RTC (via timer or arch_alarm + oneshot). The point is that oneshot interface is designed to work with absolute times, presumably with a real RTC (which does not wrap around in a long time). On up_timer_gettime from arch_alarm.c, the counter is obtained as an absolute value. Since nRF52 RTC is simply a low power timer (with a period of 512 seconds) it would be a lot more complex to offer that interface.

Reply:
It's easy to extend the period beyond 512 seconds if we monitor the overflow interrupt, but it's impossible to compensate the timer drift if we enable tickless with timer_lowerhalf_s. From our test with systick, the drift will bigger than one second after serveral hours(clock_gettime).

Reply:
> Yes, resetting and stopping the timer is the correct behavior for a oneshot timer. I expect that that all of the implementations do that.

Ok, I will try to document this since at the moment it needs to be inferred from the implementation. 
What about timer_lowerhalf_s interface? On stm32 for example, the implementation is based on the autoreload set to the timeout value. So it keeps running and the period is variable. On "advanced" STM32 timers there is a compare channel which allow to have a long and fixed autoreload (maximum value), so in this case it could be implemented differently. What is the requirement of the timer_lowerhalf_s in this regard? 

> My suspicion from what you say is that the those arch drivers are not compatible with the existing MCU architectures. They are not used by any of the existing architectures in the repository; they are used only by Xiaomi hardware. They have probably implemented an incompatible version of the one shot driver for their purposes and those arch drivers may not be usable by other architectures.
> 
> I am completely unfamiliar with those drivers. I have never used them or even looked at them. You would have to discuss this with Xiaomi to see if they are using some non-standard oneshot driver. I don't know what we should do if they are. Certainly we do not want to corrupt the semantics of the oneshot driver.

Well arch_alarm.c is not assuming something different (it manually restarts the oneshot on the handler), so it works, but since this is to be used with TICKLESS_ALARM, a timer capable of freerun + compare event should be used.